### PR TITLE
[NT] Backward support for broadcasting binary ops

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113091
* __->__ #112519
* #113031

### CODE REVIEW

File: test/test_nestedtensor.py [hunk 1]
```diff
@@ -2939,6 +2937,61 @@ def grad_test_func(a, b, c):
...
+                        nt, _ = jagged_from_list(ts, None)
+                        torch.sum(nt, dim=rd, keepdim=keepdim)
+                    continue
+
+                def grad_test_func(*ts):
+                    nt, _ = jagged_from_list(ts, None)
+                    out = torch.sum(nt, dim=rd, keepdim=keepdim)
+                    return buffer_from_jagged(out)
+
+                gradcheck(grad_test_func, inputs=(*ts,), check_batched_grad=False)
```
Comment (soulitzer) [inline test/test_nestedtensor.py:2993]: Still need test for correctness

File: torch/csrc/autograd/input_metadata.h [hunk 2]
```diff
@@ -90,39 +90,42 @@ struct InputMetadata {
...
-    // Currently NestedTensors are not expandable. If this support is added then
-    // updates to reduce_grad will be needed
-    check_nestedness_same(grad);
-    return grad.is_nested()
+    // Currently C++ NestedTensors are not expandable (but Python NestedTensors
+    // may be). If this support is added then updates to reduce_grad will be
+    // needed
+    if (is_nestedness_same(grad) &&
+        !grad.unsafeGetTensorImpl()->is_python_dispatch()) {
+      return false;
```
Comment (YuqingJ) [inline torch/csrc/autograd/input_metadata.h:116]: For the case when they both non nested and not python_dispatch, should it return at::is_expandable_to(shape_as_dim_vector(), grad.sym_sizes())?

Comment (soulitzer) [reply torch/csrc/autograd/input_metadata.h:116]: Yup, there's a typo here. This line should be: `!is_nestedness_same(grad) && !grad.unsafeGetTensorImpl()->is_python_dispatch()`.

File: torch/nested/_internal/ops.py [hunk 3]
```diff
@@ -536,6 +569,52 @@ def is_same_size_default(func, *args, **kwargs):
...
+
+    # sum_dim_IntList can produce a NT or a T depending on whether the ragged dim
+    # is summed over
+    _, new_kwargs = normalize_function(
+        func, args=args, kwargs=kwargs, normalize_to_only_use_kwargs=True
+    )
+    inp = new_kwargs.pop("input")
+    new_kwargs["dim"] = tuple(
+        canonicalize_dims(inp.dim(), s) for s in new_kwargs["dim"]
+    )
```
Comment (jbschlosser) [inline torch/nested/_internal/ops.py:586]: nit: I've been using `_wrap_jagged_dim()` for canonicalization + conversion to jagged dim space (i.e. subtracting 1). maybe that can be used here for consistency as well? I think you'd need to tweak it some to avoid the restrictive dim check under certain cases.

File: torch/nested/_internal/ops.py [hunk 4]
```diff
@@ -78,6 +78,37 @@ def raggedness_matches(nt, size):
...
+    # If unsqueezing on the 0th dim becomes supported, we would unsqueeze
+    # at step (4).
+    t_shape = t.shape
+    extra = 0
+    for s in t_shape:
+        if s == 1:
+            if t.dim() > nt.dim():
+                extra += 1
+            t = t.squeeze(0)
+    return t, extra
```
Comment (jbschlosser) [inline torch/nested/_internal/ops.py:109]: does this all work within torch.compile without unwanted guards?

File: torch/nested/_internal/ops.py [hunk 5]
```diff
@@ -161,15 +192,17 @@ def jagged_binary_pointwise(func, *args, **kwargs):
...
-        arg2 = b._values if not a_is_nt else b
-        return NestedTensor(func(arg1, arg2, *args[2:], **kwargs), **extracted_kwargs)
+    # ex: (B, j0, ?, ?) + (?) -> (B, j0, ?, ?)
+    # ex: (B, j0, ?, ?) + (?, ?) -> (B, j0, ?, ?)
+    # ex: (B, j0, ?, ?) + (1, 1, ?, ?) -> (B, j0, ?, ?)
+    nt, t = (a, b) if a_is_nt else (b, a)
+    # See Note: [ Squeezing leading ones ]
+    t_squeezed, extra = squeeze_leading_ones_get_extra(t, nt)
+    if extra != 0:
+        raise NotImplementedError()
```
Comment (jbschlosser) [inline torch/nested/_internal/ops.py:202]: nit: more informative error message here

File: torch/nested/_internal/ops.py [hunk 6]
```diff
@@ -161,15 +192,17 @@ def jagged_binary_pointwise(func, *args, **kwargs):
...
-    if (a_is_nt and a.dim() > b.dim()) or (not a_is_nt and b.dim() > a.dim()):
-        arg1 = a._values if a_is_nt else a
-        arg2 = b._values if not a_is_nt else b
-        return NestedTensor(func(arg1, arg2, *args[2:], **kwargs), **extracted_kwargs)
+    # ex: (B, j0, ?, ?) + (?) -> (B, j0, ?, ?)
+    # ex: (B, j0, ?, ?) + (?, ?) -> (B, j0, ?, ?)
+    # ex: (B, j0, ?, ?) + (1, 1, ?, ?) -> (B, j0, ?, ?)
+    nt, t = (a, b) if a_is_nt else (b, a)
+    # See Note: [ Squeezing leading ones ]
+    t_squeezed, extra = squeeze_leading_ones_get_extra(t, nt)
```
Comment (jbschlosser) [inline torch/nested/_internal/ops.py:200]: dumb Q: can you explain a bit more the purpose behind this change? I'm unclear as to what it's solving in the context of this PR

File: test/test_nestedtensor.py [hunk 7]
```diff
@@ -2844,23 +2848,21 @@ def grad_test_func(a, b, c):
...
     def _get_example_tensor_lists(self):
         return [
             # (B, *, D) with B=4
-            [
-                torch.randn(2, 5),
-                torch.randn(3, 5),
-                torch.randn(4, 5),
-                torch.randn(6, 5)
-            ],
+            self._get_list_for_jagged_tensor(((2, 3, 4, 6), 5)),
```
Comment (jbschlosser) [inline test/test_nestedtensor.py:2863]: is this going to work with the changes in #112304 that give `requires_grad=True` to some of the elements inside?

File: test/test_nestedtensor.py [hunk 8]
```diff
@@ -2939,6 +2941,82 @@ def grad_test_func(a, b, c):
...
+        # (B, j0, ?, ?) + (?, ?) -> (B, j0, ?, ?)
+        # (B, j0, ?, ?) + (1, ?, ?) -> (B, j0, ?, ?)
+        # Unsupported: (B, j0, ?, ?) + (1, 1, 1, ?, ?) -> (1, B, j0, ?, ?)
+        t_sizes = (
+            (4,),
+            (1, 4),
+            (3, 1),
+            (1, 3, 1),
+            (1, 1, 1, 4),
+            # (1, 1, 1, 1, 4), (unsupported todya)
```
Comment (jbschlosser) [inline test/test_nestedtensor.py:2957]:
```suggestion
            # (1, 1, 1, 1, 4), (unsupported today)
```

File: test/test_nestedtensor.py [hunk 9]
```diff
@@ -2939,6 +2941,82 @@ def grad_test_func(a, b, c):
...
+            # dims, expected shape, expected keepdim shape
+            # j0 is represented as None
+            ((0, 1), (3, 4), (1, 1, 3, 4)),
+            ((1, 2), None, None),
+            ((2, 3), (3, None), (3, None, 1, 1)),
+            ((0, 1, 3), (3,), (1, 1, 3, 1)),
+            ((0, 1, 2), (4,), (1, 1, 1, 4)),
+            ((0, 1, 2, 3), tuple(), (1, 1, 1, 1)),
+        )
+        for keepdim in (True, False):
```
Comment (jbschlosser) [inline test/test_nestedtensor.py:2984]: possibly `@parametrize("keepdim", [False, True])` instead of manually iterating here?

**Review by jbschlosser — COMMENTED** [review]
Thanks for the PR! got some minor comments and questions

File: torch/nested/_internal/ops.py (hunk 6 above)
Comment (soulitzer) [reply torch/nested/_internal/ops.py:200]: Previously we were too relaxed when it comes to checking whether the pre-existing broadcasting logic is valid for the given NTs. e.g. someone could try `JT: [B, *, D] + T: [sum(*), D]` and still produce something, even though broadcasting didn't really make sense in the first place. 

What we kind of what to do is to make sure that the JT has a dim that is two greater than that of the other tensor, so that we would have something like `(B, j0, a0, ..., an) + (b0, ..., bm) where m < n` where B and j0 are always broadcasted over uniformly (as in any value in a given batch acquires the same values) during broadcasting logic. This is important because otherwise we would fall into the unbind case below.

We want this `squeeze_leading_ones_get_extra` helper because naively doing `nt.dim() > t.dim() + 2` doesn't quite work though because the t might have leading ones in its dim that shouldn't disqualify it from the easy case.

File: torch/nested/_internal/ops.py (hunk 4 above)
Comment (soulitzer) [reply torch/nested/_internal/ops.py:109]: Hm let me check

File: torch/nested/__init__.py [hunk 10]
```diff
@@ -157,7 +157,8 @@ def nested_tensor(tensor_list, *, dtype=None, layout=None, device=None, requires
     elif layout == torch.jagged:
         from torch.nested._internal.nested_tensor import jagged_from_list
 
-        nt, _ = jagged_from_list(tensor_list, offsets=None, device=device, dtype=dtype)
+        with torch.no_grad():
```
[the remark this answers is not in the document]
Comment (soulitzer) [reply torch/nested/__init__.py:160]: I think they are equivalent, but I can update that one to use `torch.no_grad()` too though just for consistency.

File: test/test_nestedtensor.py (hunk 7 above)
Comment (soulitzer) [reply test/test_nestedtensor.py:2863]: Good catch, yeah looks like this would need to be updated.

File: torch/nested/_internal/nested_tensor.py [hunk 11]
```diff
@@ -97,6 +97,7 @@ def __init__(self, values, offsets, *, ragged_size=None, **kwargs):
             )
         self._values = values
         self._offsets = offsets
+        self.requires_grad = kwargs.get("requires_grad", False)
```
[the remark this answers is not in the document]
Comment (soulitzer) [reply torch/nested/_internal/nested_tensor.py:100]: Yup, I have no idea why I didn't that 🤦

File: torch/nested/_internal/ops.py (hunk 3 above)
Comment (soulitzer) [reply torch/nested/_internal/ops.py:586]: Good idea, I refactored this a little.

File: test/test_nestedtensor.py (hunk 7 above)
Comment (soulitzer) [reply test/test_nestedtensor.py:2863]: I ended up just changing it back.

File: torch/nested/_internal/ops.py (hunk 4 above)
Comment (soulitzer) [reply torch/nested/_internal/ops.py:127]: I do think we do want the extra guards here.

File: torch/nested/_internal/ops.py (hunk 6 above)
Comment (jbschlosser) [reply torch/nested/_internal/ops.py:218]: cool, thanks for the explanation :)

File: torch/nested/_internal/ops.py (hunk 4 above)
Comment (jbschlosser) [reply torch/nested/_internal/ops.py:127]: hm okay, which guards show up?

Comment (soulitzer) [reply torch/nested/_internal/ops.py:133]: > I do think we do want the extra guards here.

Whoops, this is wrong actually. What is happening is that there ARE always guards testing whether inputs given are zero/one or not (due to zero-one specialization) which is indeed what we want, but also, as a result of that there are also no "extra" guards from this `s == 1` test here.

File: torch/nested/_internal/ops.py [hunk 12]
```diff
@@ -536,6 +569,52 @@ def is_same_size_default(func, *args, **kwargs):
...
 
 
+@register_jagged_func(
+    torch.ops.aten.sum.dim_IntList, "self: jt, dim: any?, keepdim: any?, dtype: any?"
+)
+def sum_dim_IntList(func, *args, **kwargs):
+    from torch._prims_common import canonicalize_dims
+
+    # sum_dim_IntList can produce a NT or a T depending on whether the ragged dim
+    # is summed over
```
[the remark this answers is not in the document]
Comment (soulitzer) [reply torch/nested/_internal/ops.py]: I think returning plain T has more known benefits, and it's not clear if returning NT today would give us more flexibility wrt bc either. If we return NT today and make users to grab values explicitly, that would no longer work if we choose to return T later.

Pros:
- avoid having additional state on NT tracking whether we are a dense NT or not
- less additional logic within autograd to handle the conversion from dense NT to dense
- users would not have to worry about having to manually converting the dense NT back to dense to avoid subclass overhead

Cons:
- some ops on NT unexpectedly(?) return non-NTs. I'm not actually sure how problematic it is. On the NT front, as long as NTs can freely promote, doing any NT related ops should not be problematic on this dense T.

[the remark this answers is not in the document]
Comment (soulitzer) [reply torch/nested/_internal/ops.py]: cc @albanD

File: torch/csrc/autograd/input_metadata.h [hunk 13]
```diff
@@ -90,39 +90,42 @@ struct InputMetadata {
...
   }
 
   at::Tensor reduce_grad(at::Tensor& grad) const {
     // Currently reduce_grad is only called if is_expandable_to_shape returns
-    // true For nested tensors this always returns False, so this check
-    // shouldn't fail
-    TORCH_INTERNAL_ASSERT(!grad.is_nested() && !is_nested_)
+    // true. For C++ nested tensors this always returns False, so this check
+    // shouldn't fail.
+    TORCH_INTERNAL_ASSERT(!(
```
Comment (albanD) [inline torch/csrc/autograd/input_metadata.h:127]: Adding so much nested-specific code here is less than ideal...
It would be great to have nicer functions for each of these checks rather than having to check dispatch keys inline...

Comment (soulitzer) [reply torch/csrc/autograd/input_metadata.h:127]: Agreed. Going to factor this out. I will put the function in this file for now until somewhere else needs it.

File: torch/nested/_internal/ops.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (soulitzer) [reply torch/nested/_internal/ops.py]: From offline discussion @albanD thinks this seems fine as long as there are no silent correctness issues.

[the remark this answers is not in the document]
Comment (jbschlosser) [reply torch/nested/_internal/ops.py]: Cool I'm good with it then :)

File: torch/nested/_internal/ops.py [hunk 14]
```diff
@@ -79,6 +112,37 @@ def raggedness_matches(nt, size):
...
+    #
+    # If unsqueezing on the 0th dim becomes supported, we would unsqueeze
+    # at step (4).
+    t_shape = t.shape
+    extra = 0
+    for s in t_shape:
+        if s == 1:
+            if t.dim() > nt.dim():
+                extra += 1
+            t = t.squeeze(0)
```
Comment (jbschlosser) [inline torch/nested/_internal/ops.py:142]: I'm reading this and it's hard to tell if it's correct. Looks like it depends on `t.squeeze(0)` being a no-op when `t.shape[0] > 1`. Also `1`s we see in the middle of t's shape shouldn't contribute to `extra`, right? Is it possible to simplify this and still get the correct guard behavior? For example:

```python
num_leading_ones = 0
while t.shape[0] == 1:
    num_leading_ones += 1
    t = t.squeeze(0)
return t, num_leading_ones
```

File: torch/nested/_internal/ops.py (hunk 6 above)
Comment (jbschlosser) [reply torch/nested/_internal/ops.py:235]: Note that if we see `t.dim() > nt.dim()`, we know we're in an unsupported case, even if `t` is all ones. If we bail out early for this, it might be a little more readable.

**Review by jbschlosser — APPROVED** [review]
LGTM thanks!

one nit though: I do think the NT, T broadcasting handling could be a bit more straightforward

File: torch/nested/_internal/ops.py (hunk 14 above)
Comment (soulitzer) [reply torch/nested/_internal/ops.py]: Thanks, good catch, was missing the else-break here, and this simplification makes things much cleaner!
